ambry-router/src/main/java/com/github/ambry/router/CompressionService.java

```diff
@@ -0,0 +1,350 @@
+      compressionMetrics.compressSkipRate.mark();
+      compressionMetrics.compressSkipContentTypeFiltering.inc();
+      return false;
+    }
+
+    // For all other case, compression should be enabled.
+    return true;
+  }
+
+  /**
+   * Apply compression to a chunk buffer if compression is enabled and content-type is compressible.
+   * It applies compression config to determine whether to accept or reject the compressed result.
+   * For example, if the compressed data size is smaller than original data by certain % (exceeds threshold),
+   * the compressed data is accepted and return compressed data; otherwise, compression is discarded.
+   * This method emits metrics along the process.  It returns null if failed instead of throwing exceptions.
+   * This method is typically called for the first chunk to determine whether to compress the entire blob or not.
+   * Subsequent data chunks are compressed using compressChunk().
+   *
+   * @param chunkBuffer The PutChunk buffer to compress.
+   * @param isFullChunk Whether this is a full size chunk (4MB) or smaller.
+   * @return Returns a new compressed buffer.  It returns null is no compression is applied.
+   */
+  public ByteBuf compressChunk(ByteBuf chunkBuffer, boolean isFullChunk) {
+    Objects.requireNonNull(chunkBuffer, "blobProperties cannot be null.");
+
+    // Check the blob size.  If it's too small, do not compress.
+    int sourceDataSize = chunkBuffer.readableBytes();
+    if (sourceDataSize < compressionConfig.minimalSourceDataSizeInBytes) {
+      logger.info("No compression applied because source data size " + sourceDataSize + " is smaller than minimal size "
+          + compressionConfig.minimalSourceDataSizeInBytes);
+      // Emit metrics to count how many compressions skipped due to small size.
+      compressionMetrics.compressSkipRate.mark();
+      compressionMetrics.compressSkipSizeTooSmall.inc();
+      return null;
+    }
+
+    // Convert ByteBuf to byte[] before applying compression.
+    byte[] sourceBuffer;
+    int sourceBufferOffset;
+    try {
+      Triple<byte[], Integer, Integer> bufferInfo = convertByteBufToByteArray(chunkBuffer);
+      sourceBuffer = bufferInfo.getLeft();
+      sourceBufferOffset = bufferInfo.getMiddle();
+      sourceDataSize = bufferInfo.getRight();
+    } catch (Exception ex) {
+      logger.error(String.format("Cannot compress source data because failed to convert ByteBuf to byte[]."
+              + "chunkBuffer.hasArray() = %s, chunkBuffer.readableBytes = %d, chunkBuffer.arrayOffset = %d",
+          chunkBuffer.hasArray(), chunkBuffer.readableBytes(), chunkBuffer.arrayOffset()), ex);
+
+      // Emit metrics for count how many compressions skipped due to buffer conversion error.
+      compressionMetrics.compressErrorRate.mark();
+      compressionMetrics.compressErrorBufferConversion.inc();
+      return null;
+    }
+
+    // Get default compressor.  If config missing, fallback to Zstandard.
```

Comment:
nit: maybe add the logic to get the default compressor in a separate method.

Reply:
Getting the default compressor isn't called by another method (not shared).  Also, it's not big, just 3 lines.  Breaking it up to another method shortens this method creates more methods which doesn't reduce code side.  I don't think it makes a difference by breaking it out, unless it has readability issue.